### PR TITLE
Brush Classes refactoring

### PR DESCRIPTION
**Pull request by FlorianLudwig**
In preparation to implement more different line and fill styles we would like to refactor GraphicsData to use a new class to store brush details. See #1567

### CODE REVIEW

**Comment by drkibitz** [comment]
My initial reaction is :+1:, just with a couple questions...

- What's the reason to use a getter for `canvasBrush`? Would a possible micro optimization (similar to caching the value), be setting `_canvasBrush` in a `tint` setter?

- Would using `"rgba(x,x,x,x)"` format for `context.strokeStyle` and `context.fillStyle` simplify anything here for _canvasBrush?

**Comment by MishaAmashukeli** [comment]
Good points, thanks.
The getter for canvasBrush was there for lazy evaluation, because I did not want anything canvas-related to happen when the renderer is webgl-renderer. But for SolidBrush that does not really make sense, so I changed that.
I initially considered creating canvas and webgl related subclasses  but then decided that would create too many classes and too deep hierarchy. 

As for rgba vs setting context.globalAlpha - I'll try that out.

**Comment by MishaAmashukeli** [comment]
Setting globalAlpha does not seem to have any performance issues. Using RGBA would force us to regenerate the value on each frame if we're animating the graphics alpha, so I think we don't need it.

**Comment by MishaAmashukeli** [comment]
tint arg is not used by SolidBrush, but it will be used by the TextureBrush:

    TextureBrush.prototype.setTint = function (tint)
    {
        var texture = (tint === 0xFFFFFF)
            ? this.texture.baseTexture.source
            : CanvasTinter.getTintedTexture({texture: this.texture}, tint);
    ......

We can remove tint arg in SolidBrush.setTint and add a condition in CanvasGraphics.updateGraphicsTint to pass only tintR/tintG/tintB if the brush is SolidBrush, but imho this is not necessary... 
As far as jshint is concerned, this is not an error.

File: src/core/graphics/brushes/SolidBrush.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+/*
+ * Generates _canvasBrush based on the tint color
+ *
+ * @param tint {Number} the tint color
+ * @param tintR {Number} the R component of the tint color
+ * @param tintG {Number} the G component of the tint color
+ * @param tintB {Number} the B component of the tint color
+ *
+ */
+SolidBrush.prototype.setTint = function (tint, tintR, tintG, tintB)
```
Comment (englercj) [inline src/core/graphics/brushes/SolidBrush.js]: This seems wrong, and definitely does not follow the pattern in the rest of pixi.

Most likely this should just be a tint color param that uses `utils.hex2rgb()`. Having both params doesn't really make sense.

**Comment by MishaAmashukeli** [comment]
Ok, changed it.

**Comment by englercj** [comment]
I like the direction of this, @GoodBoyDigital can you review as well?

**Comment by FlorianLudwig** [comment]
@englercj @GoodBoyDigital I would really like to see this merged, is there something left I could help with to move this forward?

**Comment by stbaer** [comment]
+1 for getting this in.

**Comment by GoodBoyDigital** [comment]
Hi @FlorianLudwig , apologies for not getting on this sooner. Busy times at the moment. I will get round to checking this out soon. In the mean time feel free to push forward with your branch as I'm keen to see where this goes!

**Comment by englercj** [comment]
@FlorianLudwig Have you worked on this any more? I'm interested in merging this in, just curious if there has been more movement on it and if we should start a new PR that is off the latest develop with your latest brush-related code.

**Comment by FlorianLudwig** [comment]
@englercj The intention was to enable bitmapfill and radiantfill in the long run.  The project where we needed those was solved with what is already available in pixi, so we did not pursue it any further.  But I still believe those APIs would be useful within pixi.  I will look if I can schedule someone to take a look at rebasing it and finishing the bitmapfill work.

**Comment by englercj** [comment]
Anyway this can rebased onto the lastest `dev` branch?

```
git checkout brushes
git rebase dev
git push -f origin brushes
```

**Comment by FlorianLudwig** [comment]
@englercj done

**Comment by Kran67** [comment]
@FlorianLudwig Can you make an example to use gradient brush in graphics shape?

In this exemple : http://pixijs.github.io/examples/index.html?s=demos&f=graphics-demo.js&title=Graphics

Thanks

**Comment by englercj** [comment]
This is something I want to do in v5, but I don't think it will happen in v4. I'll keep this open as a reminder and reference implementation though. Thanks!

**Comment by bigtimebuddy** [comment]
I'm closing this since @englercj added #3123 which references this.

**Pull request closed without merging**